tools/src/wyvern/tools/parsing/coreparser/arch/DeclCheckVisitor.java

```diff
@@ -0,0 +1,224 @@
+                }
+                for (String reqs : compType.getReqs().values()) {
+                    for (String innerfullport : fullports) {
+                        String innercomponent = innerfullport.substring(0, innerfullport.indexOf("."));
+                        ASTComponentDecl innercompObj = components.get(innercomponent);
+                        ASTComponentTypeDecl innercompType = componentTypes.get(innercompObj.getType());
+                        if (!nodes.containsKey(innercomponent)) {
+                            DependencyGraphNode innercompNode = new DependencyGraphNode(innercompObj);
+                            nodes.put(innercomponent, innercompNode);
+                        }
+                        DependencyGraphNode innercompNode = nodes.get(innercomponent);
+                        for (String provs : innercompType.getProvs().values()) {
+                            if (provs.equals(reqs)) {
+                                compNode.addRequires(innercompNode);
+                                innercompNode.addProvides(compNode);
+                                nodes.put(innercomponent, innercompNode);
+                                nodes.put(component, compNode);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // Apply topological sort to the graph!
+        for (DependencyGraphNode m : noReqNodes) {
+            ordered.add(m.getComponentDecl());
+            for (DependencyGraphNode n : m.getProvides()) {
+                n.getProvides().remove(m);
+                if (n.getProvides().size() == 0) {
+                    ordered.add(n.getComponentDecl());
+                }
+            }
+        }
+        return ordered;
+    }
+
+    public HashMap<String, ASTComponentTypeDecl> getComponentTypes() {
+        return componentTypes;
+    }
+
+    public HashSet<String> getConnectorTypes() {
+        return connectorTypes;
+    }
+
+    public HashMap<String, ASTComponentDecl> getComponents() {
+        return components;
+    }
+
+    public HashMap<String, String> getConnectors() {
+        return connectors;
+    }
+
+    public HashMap<String, String> getEntrypoints() {
+        return entrypoints;
+    }
+
+    public HashMap<String, HashSet<String>> getAttachments() {
+        return attachments;
+    }
+
+    public HashMap<String, ASTPortDecl> getPortDecls() {
+        return portdecls;
+    }
+
+    public DeclCheckVisitor(InterpreterState state) {
+        super();
+        this.state = state;
+    }
+
+    public Object visit(ASTAttachmentDecl node, Object data) {
+        String connector = node.getConnector();
+        HashSet<String> ports = node.getAllPorts();
+        if (!connectors.containsKey(connector)) {
+            // connector not declared
+            ToolError.reportError(ErrorMessage.MEMBER_NOT_DECLARED,
+                    node.getLocation(), "connector", connector);
+        }
```

Comment:
There is a subtle bug here. The grammar allows attachments to be declared before connectors. If that happens, 'connectors' would be empty and an error would be reported even though the connector might actually be declared. If you have time to fix it, please fix it. Otherwise, add a FIXME comment here noting this issue.